### PR TITLE
feat(1415): Add logic to handle webhook events from external repos

### PR DESCRIPTION
**Pull request by supra08**
## Context

Fixes screwdriver-cd/screwdriver#1415

## Objective

This PR adds logic to webhook endpoint to handle notifications from external repos. First it filters the triggered pipelines with the checkoutUrl from the incoming request. Here it utilizes the newly added `contains` operator for querying. After that it is checked whether the request is from an external repo or not and based on that sends a flag to the EventFactory module to handle them likewise.

## References

screwdriver-cd/screwdriver#1415

## License
I confirm that this contribution is made under a BSD license and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: plugins/webhooks/index.js [hunk 1]
```diff
@@ -279,9 +282,12 @@ async function triggeredPipelines(
     const scmBranch = `${splitUri[0]}:${splitUri[1]}:${splitUri[2]}`;
     const scmRepoId = `${splitUri[0]}:${splitUri[1]}`;
     const listConfig = { search: { field: 'scmUri', keyword: `${scmRepoId}:%` } };
+    const externalRepoSearchConfig = { params: { contains: { subscribedScmUrls: scmUri } } };
 
     const pipelines = await pipelineFactory.list(listConfig);
 
+    const pipelinesWithSubscribedExtRepos = await pipelineFactory.list(externalRepoSearchConfig);
```
Comment (jithine) [inline plugins/webhooks/index.js:289]: we can call this `pipelinesWithSubscribedRepos`

File: plugins/webhooks/index.js [hunk 2]
```diff
@@ -385,6 +394,15 @@ async function createPREvents(options, request) {
                 chainPR
             });
 
+            let subscribedEvent = false;
```
Comment (jithine) [inline plugins/webhooks/index.js:397]: don't you need similar changed in `createEvents` function ? This function is called only for PR events

Comment (supra08) [reply plugins/webhooks/index.js:397]: yeah they are coming up as further additions. This was just a work in progress.

File: plugins/webhooks/index.js [hunk 3]
```diff
@@ -1,3 +1,6 @@
+/* eslint-disable max-lines-per-function */
+/* eslint-disable max-params */
```
Comment (tkyi) [inline plugins/webhooks/index.js:2]: should remove

File: plugins/webhooks/index.js [hunk 4]
```diff
@@ -250,6 +253,14 @@ function getSkipMessageAndChainPR({ pipeline, prSource, restrictPR, chainPR }) {
     return result;
 }
 
+const uriTrimmer = uri => {
```
Comment (tkyi) [inline plugins/webhooks/index.js:256]: jsdocs?

File: plugins/webhooks/index.js [hunk 5]
```diff
@@ -366,16 +386,41 @@ async function createPREvents(options, request) {
...
+            let eventConfig = {};
 
-            try {
-                configPipelineSha = await pipelineFactory.scm.getCommitSha(scmConfig);
-            } catch (err) {
-                if (err.status >= 500) {
-                    throw err;
-                } else {
-                    logger.info(`skip create event for branch: ${b}`);
+            // Check is the webhook event is from a subscribed repo and
```
Comment (tkyi) [inline plugins/webhooks/index.js:392]:
```suggestion
            // Check if the webhook event is from a subscribed repo and
```

File: plugins/webhooks/index.js [hunk 6]
```diff
@@ -399,19 +446,47 @@ async function createPREvents(options, request) {
...
 
-            if (skipMessage) {
-                eventConfig.skipMessage = skipMessage;
-            }
-
             if (b === branch) {
                 eventConfig.startFrom = '~pr';
             }
 
+            // Check is the webhook event is from a subscribed repo and
```
Comment (tkyi) [inline plugins/webhooks/index.js:458]:
```suggestion
            // Check if the webhook event is from a subscribed repo and
```

File: plugins/webhooks/index.js [hunk 7]
```diff
@@ -350,14 +368,19 @@ async function createPREvents(options, request) {
...
-        chainPR
+        chainPR,
+        ref,
+        releaseName,
+        meta
     } = options;
     const { scm } = request.server.app.pipelineFactory;
     const { eventFactory } = request.server.app;
     const { pipelineFactory } = request.server.app;
+    const { userFactory } = request.server.app;
```
Comment (tkyi) [inline plugins/webhooks/index.js:379]: can condense with previous two lines

File: plugins/webhooks/index.js [hunk 8]
```diff
@@ -399,19 +447,47 @@ async function createPREvents(options, request) {
...
-                eventConfig.skipMessage = skipMessage;
-            }
-
             if (b === branch) {
                 eventConfig.startFrom = '~pr';
             }
 
+            // Check if the webhook event is from a subscribed repo and
+            // set the jobs entrypoint from ~startfrom
+            // For subscribed PR event, it should be mimiced as a commit
```
Comment (tkyi) [inline plugins/webhooks/index.js:461]: what does this mean? should this be `type: 'pr'`?

Comment (jithine) [reply plugins/webhooks/index.js:461]: pr event happened on the subscribed git repo. But for the subscriber it's similar to running a `commit` event

**Review by jithine — APPROVED** [review]

**Review by tkyi — APPROVED** [review]

**Pull request merged**